src/circuit/gadget/utilities/cond_swap.rs

```diff
@@ -0,0 +1,319 @@
+                        value: x_swapped,
+                    }
+                };
+
+                // Conditionally swap y
+                let y_swapped = {
+                    let y_swapped = x
+                        .value
+                        .zip(y.value)
+                        .zip(swap)
+                        .map(|((x, y), swap)| if swap { x } else { y });
+                    let y_swapped_cell = region.assign_advice(
+                        || "y_swapped",
+                        config.y_swapped,
+                        0,
+                        || y_swapped.ok_or(Error::SynthesisError),
+                    )?;
+                    CellValue {
+                        cell: y_swapped_cell,
+                        value: y_swapped,
+                    }
+                };
+
+                // Return swapped pair
+                Ok((x_swapped, y_swapped))
+            },
+        )
+    }
+}
+
+impl<F: FieldExt> CondSwapChip<F> {
+    /// Configures this chip for use in a circuit.
+    pub fn configure(
+        meta: &mut ConstraintSystem<F>,
+        advices: [Column<Advice>; 5],
+        perm: Permutation,
+    ) -> CondSwapConfig {
+        let q_swap = meta.selector();
+
+        let config = CondSwapConfig {
+            q_swap,
+            x: advices[0],
+            y: advices[1],
+            x_swapped: advices[2],
+            y_swapped: advices[3],
+            swap: advices[4],
+            perm,
+        };
+
+        // TODO: optimise shape of gate for Merkle path validation
+
+        meta.create_gate("x' = y ⋅ swap + x ⋅ (1-swap)", |meta| {
+            let q_swap = meta.query_selector(q_swap, Rotation::cur());
+
+            let x = meta.query_advice(config.x, Rotation::cur());
+            let y = meta.query_advice(config.y, Rotation::cur());
+            let x_swapped = meta.query_advice(config.x_swapped, Rotation::cur());
+            let y_swapped = meta.query_advice(config.y_swapped, Rotation::cur());
+            let swap = meta.query_advice(config.swap, Rotation::cur());
+
+            let one = Expression::Constant(F::one());
+
+            // x_swapped - y ⋅ swap - x ⋅ (1-swap) = 0
+            // This checks that `x_swapped` is equal to `y` when `swap` is set,
+            // but remains as `x` when `swap` is not set.
+            let x_check =
+                x_swapped - y.clone() * swap.clone() - x.clone() * (one.clone() - swap.clone());
+
+            // y_swapped - x ⋅ swap - y ⋅ (1-swap) = 0
+            // This checks that `y_swapped` is equal to `x` when `swap` is set,
+            // but remains as `y` when `swap` is not set.
+            let y_check = y_swapped - x * swap.clone() - y * (one.clone() - swap.clone());
+
+            // Check `swap` is boolean.
+            let bool_check = swap.clone() * (one - swap);
+
+            [x_check, y_check, bool_check]
+                .iter()
+                .map(|poly| q_swap.clone() * poly.clone())
+                .collect()
```

Comment:
I think this is a useful-enough pattern (returning several constraints that are all controlled by a single selector in the simplest way) that we could abstract it into `halo2`. Unfortunately we can't do that with `From` because both `Vec` and tuples are foreign types, but we could use a macro.

In the meantime, this can be slightly more efficient (with a `use std::array`):
```suggestion
            array::IntoIter::new([x_check, y_check, bool_check])
                .map(|poly| q_swap.clone() * poly)
                .collect()
```